### ISSUE TITLE
Keras should offer a Callback to Pick the Best model weights

### ISSUE THREAD

**Issue by tim-janik** [keras-team/keras#18046]
**System information**.

Tensorflow: 2.12.0
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

I tried to preserve the best weights of a model.fit() run with `tf.keras.callbacks.EarlyStopping (patience=99e99, restore_best_weights=True)`, I didn't want early stopping (cause configuring that felt too complicated) but preserve the weights of the best epoch nevertheless. This didn't work b/c EarlyStopping only restores weights after its patience phase.

So IMHO, EarlyStopping should be changed to restore the best weights unconditionally if  restore_best_weights=True, or (since the former may introduce incompatible behavior) a Keras callback can be provided that achieves always picking the best weights.

**Will this change the current api? How?**

Potentially, I'm not sure if users are relying on `EarlyStopping(restore_best_weights=True)` to *not* restore the weights if the model.fit() run stopped for a reason other than EarlyStopping running out of patience...

**Who will benefit from this feature?**

All users of Keras that do not manage to have their loss monotonically decreasing (or accuracy monotonically increasing) will have the ability to keep the best model weights that they managed to calculate in the past but that are currently thrown away.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes, if needed
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

Short of EarlyStopping being fixed, here is an example of what I'm suggesting, based on the EarlyStopping implementation:

```py
from keras.utils import io_utils
from tensorflow.python.platform import tf_logging
class PickBest (tf.keras.callbacks.Callback):
  """Pick (restore) the best model weights after training stops."""
  def __init__(self, monitor = 'val_loss', min_delta = 0, mode = 'auto', verbose = 1):
    super (PickBest, self).__init__()
    self.monitor = monitor
    self.min_delta = abs (min_delta)
    self.verbose = verbose
    self.best_weights = None
    if (mode == 'max' or self.monitor.endswith ('auc') or
        self.monitor.endswith ('acc') or self.monitor.endswith ('accuracy')):
      self.monitor_op = np.greater
    else:
      self.monitor_op = np.less
      if mode not in ('min', 'auto'):
        tf_logging.warning ('PickBest: unknown mode: %s' % mode)
    if self.monitor_op == np.greater:
      self.min_delta *= 1
    else:
      self.min_delta *= -1        
  def on_train_begin (self, logs = None):
    # Cleanup for instances reuse
    self.best = np.Inf if self.monitor_op == np.less else -np.Inf
    self.best_weights = None
    self.best_epoch = 0
  def on_epoch_end (self, epoch, logs = None):
    current = self.get_monitor_value (logs)
    if current is None:
      return
    if self.best_weights == None or self._is_improvement (current, self.best):
      self.best_epoch = epoch
      self.best = current
      self.best_weights = self.model.get_weights()
  def on_train_end (self, logs=None):
    if self.best_weights is not None:
      self.model.set_weights (self.best_weights)
      if self.verbose > 0:
        io_utils.print_msg ('PickBest: Using best model weights: ' +
                            f'epoch={self.best_epoch} ' +
                            f'{self.monitor}={self.best}')
  def get_monitor_value (self, logs):
    logs = logs or {}
    monitor_value = logs.get (self.monitor)
    if monitor_value is None:
      tf_logging.warning (f'PickBest conditioned on metric `{self.monitor}` ' +
                          'which is not available. Available metrics are: ' +
                          ','.join (list (logs.keys())))
    return monitor_value
  def _is_improvement(self, monitor_value, reference_value):
    return self.monitor_op (monitor_value - self.min_delta, reference_value)
```

Example output:
```sh
Epoch 200/200
4/4 [==============================] - 1s 195ms/step - loss: 0.5467 - accuracy: 0.9826 - val_loss: 0.6435 - val_accuracy: 0.8731 - learning_rate: 9.9402e-04 - lr: 9.9402e-04
PickBest: Using best model weights: epoch=170 accuracy=0.9981288313865662
```

**Comment by Frightera** [comment]
@tim-janik 

Actually there is [tf.keras.callbacks.ModelCheckpoint](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint) which can be configured to save the model weights if there is an improvement on the selected metric. The only thing it does not restore the weights so they should be loaded manually.

Maybe you can extend this callback to support this instead of writing one from the scratch.

**Comment by tim-janik** [comment]
> @tim-janik
> 
> Actually there is [tf.keras.callbacks.ModelCheckpoint](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint) which can be configured to save the model weights if there is an improvement on the selected metric. The only thing it does not restore the weights so they should be loaded manually.
> 
> Maybe you can extend this callback to support this instead of writing one from the scratch.

That may be possible.
The point of this bug report is whether EarlyStopping can be changed to restore weights, regardless of the patience phase. 
In case the answer to that is a solid "no", other options can be explored in details. So I'm waiting on input about the EarlyStopping behaviour.

**Comment by rchao** [comment]
Thanks for the question - by design the `EarlyStopping` callback only restores the weights once the training is stopped. Otherwise, I'm not quite sure how it would make sense that the model gets restored to the best weights, while the training is still going on (you would end up with a loop that never ends, I suppose?)

Using the `ModelCheckpoint` callback with the option to only save the best weights, and after the end of your training, restoring the weights manually, seems the best solution here.

**Comment by sachinprasadhs** [comment]
Hi @tim-janik , Please refer the above comment and the documentation here about `ModelCheckpoint` for more details https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint
